as/as.1
.\" Copyright (c) 2012 Joseph Koshy <jkoshy@users.sourceforge.net>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer
.\"    in this position and unchanged.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $Id: as.1 3642 2018-10-14 14:24:28Z jkoshy $
.\"
.Dd December 22, 2012
.Dt AS 1
.Os
.Sh NAME
.Nm as
.Nd an assembler
.Sh SYNOPSIS
.Nm
.Op Fl D
.Op Fl I Ar dir
.Op Fl J
.Op Fl K
.Op Fl L | Fl -keep-locals
.Op Fl R
.Op Fl V
.Op Fl W | Fl -no-warn
.Op Fl Z
.Op Fl a Ns Oo Ar cdghlns Oc Ns Oo Ar =filename Oc
.Op Fl g | Fl -gen-debug
.Op Fl h | Fl -help
.Op Fl march= Ns Ar cpu Ns Oo ,+ Ns Oo Ar extensions Oc Ns "..." Oc
.Op Fl mtune= Ns Ar cpu
.Op Fl n
.Op Fl o Ar obj
.Op Fl q
.Op Fl -MD Ar file
.Op Fl -defsym Ar symbol Ns = Ns Ar value
.Op Fl -fatal-warnings
.Op Fl -listing-lhs-width
.Op Fl -listing-lhs-width2
.Op Fl -listing-rhs-width
.Op Fl -listing-cont-lines
.Op Fl -statistics
.Op Fl -strip-local-absolute
.Op Fl -target-help
.Op Fl -version
.Op Fl -warn
.Op Ar target-options
.Op Ar
.Sh DESCRIPTION
The
.Nm
utility creates an ELF relocatable object from assembly language
sources.
.Pp
If multiple source files are specified,
.Nm
will read each of them in the order specified.
If no files are specified,
.Nm
will read from standard input.
.Pp
The
.Nm
utility understands the following options:
.Bl -tag -width indent
.It Fl D
Display internal debugging messages.
.It Fl I Ar dir
Add the directory named by argument
.Ar dir
to the search path used by the
.Ic ".include"
directive.
.It Fl J
Suppress warnings about overflows in signed arithmetic.
.It Fl K
Warn about alterations to difference tables if any.
.It Fl L | Fl -keep-locals
Keep local symbols in the output file.
The default is to discard local symbols.
.It Fl R
Merge the
.Li .data
and
.Li .text
sections when creating an object.
.It Fl V
Print the assembler version on standard output before assembling
the inputs.
.It Fl W | Fl -no-warn
Suppress assembler warnings.
.It Fl Z
Generate an output object even if there were errors in the input.
.It Fl a Ns Oo Ar cdghlns Oc Ns Oo Ar =filename Oc
Control generated listings.
The supported flags are:
.Bl -tag -width indent -compact
.It c
Omit listing code in false conditional paths.
.It d
Omit debugging directives.
.It g
Include the version of the assembler and other general information
in the generated listing.
.It h
Include high-level source in the listing.
.It l
Include assembly source in the listing.
.It m
Include macro expansions in the listing.
.It n
Suppress generation of the header and footer in the listing.
.It s
Include symbol information in the listing.
.It = Ns Ar filename
Set the name of the listing file to that specified by argument
.Ar filename .
If specified, this flag must be the last in the list.
.El
If option
.Fl a
is specified without additional arguments, a flag set of
.Sq Ar hls
is assumed.
.It Fl g | Fl -gen-debug
Generate debug information in DWARF format.
.It Fl h | Fl -help
Show a help message and exit.
.It Fl march Ns = Ns Ar cpu Ns Oo ,+ Ns Oo Ar extensions Oc Ns "..." Oc
Generated code for the CPU named by argument
.Ar cpu
with additional instruction set extensions named by the argument
.Ar extensions .
.It Fl mtune Ns = Ns Ar cpu
Optimize the assembled object for the CPU named by the argument
.Ar cpu .
.It Fl n
Do not optimize code alignment.
.It Fl o Ar filename
Write the assembled output to the file named by argument
.Ar filename .
.It Fl q
Suppress warnings.
.It Fl -MD Ar filename
Write dependency information in a form usable by
.Xr make 1
to the file name by argument
.Ar filename .
.It Fl -defsym Ar symbol Ns = Ns Ar value
Define symbol named by the argument
.Ar symbol
as having the value named by
.Ar value .
The argument value is an integer in one of the forms accepted
by
.Xr atoi 3 .
.It Fl -fatal-warnings
Treat all warnings as fatal.
.It Fl -listing-lhs-width Ns = Ns Ar nwords
Set the width of the output data column in a listing to the number
of machine words specified in argument
.Ar nwords .
.It Fl -listing-lhs-width2 Ns = Ns Ar nwords
Set the width of continuation lines for the data column to the number
of machine words specified in argument
.Ar nwords .
.It Fl -listing-rhs-width Ns = Ns Ar nchars
Set the maximum displayed width of an input source line to the number
of characters specified by argument
.Ar nchars .
.It Fl -listing-cont-lines Ns = Ns Ar nlines
Set the maximum number of listing lines generated by one input source
line to
.Ar nlines Ns No + Ns 1 .
.It Fl -statistics
Print statistics for the run at exit.
.It Fl -strip-local-absolute
Remove local absolute symbols from the generated output.
.It Fl -target-help
Display help for the target CPU.
.It Fl -version
Print a version identifier and exit.
.It Fl -warn
Print warnings.
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr elfcopy 1 ,
.Xr ld 1 ,
.Xr nm 1 ,
.Xr strings 1 ,
.Xr strip 1 ,
.Xr elf 5


NEWS
--
NTP 4.2.8p12 (Harlan Stenn <stenn@ntp.org>, 2018/14/09)

NOTE: this NEWS file will be undergoing more revisions.

Focus: Security, Bug fixes, enhancements.

Severity: MEDIUM

This release fixes a "hole" in the noepeer capability introduced to ntpd
in ntp-4.2.8p11, and a buffer overflow in the openhost() function used by
ntpq and ntpdc.  It also provides 26 other bugfixes, and 4 other improvements:

* [Sec 3505] Buffer overflow in the openhost() call of ntpq and ntpdc.

* [Sec 3012] Fix a hole in the new "noepeer" processing.

* Bug Fixes:
 [Bug 3521] Fix a logic bug in the INVALIDNAK checks.  <stenn@ntp.org>
 [Bug 3509] Add support for running as non-root on FreeBSD, Darwin,
            other TrustedBSD platforms
 - applied patch by Ian Lepore <perlinger@ntp.org>
 [Bug 3506] Service Control Manager interacts poorly with NTPD <perlinger@ntp.org>
 - changed interaction with SCM to signal pending startup
 [Bug 3486] Buffer overflow in ntpq/ntpq.c:tstflags() <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3485] Undefined sockaddr used in error messages in ntp_config.c <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3484] ntpq response from ntpd is incorrect when REFID is null <perlinger@ntp.org>
 - rework of ntpq 'nextvar()' key/value parsing
 [Bug 3482] Fixes for compilation warnings (ntp_io.c & ntpq-subs.c) <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods)
 [Bug 3480] Refclock sample filter not cleared on clock STEP <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3479] ctl_putrefid() allows unsafe characters through to ntpq <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods)
 [Bug 3476]ctl_putstr() sends empty unquoted string [...] <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods); not sure if that's bug or feature, though
 [Bug 3475] modify prettydate() to suppress output of zero time <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3474] Missing pmode in mode7 peer info response <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3471] Check for openssl/[ch]mac.h.  HStenn.
 - add #define ENABLE_CMAC support in configure.  HStenn.
 [Bug 3470] ntpd4.2.8p11 fails to compile without OpenSSL <perlinger@ntp.org>
 [Bug 3469] Incomplete string compare [...] in is_refclk_addr <perlinger@ntp.org>
 - patch by Stephen Friedl
 [Bug 3467] Potential memory fault in ntpq [...] <perlinger@ntp.org>
 - fixed IO redirection and CTRL-C handling in ntq and ntpdc
 [Bug 3465] Default TTL values cannot be used <perlinger@ntp.org>
 [Bug 3461] refclock_shm.c: clear error status on clock recovery <perlinger@ntp.org>
 - initial patch by Hal Murray; also fixed refclock_report() trouble
 [Bug 3460] Fix typo in ntpq.texi, reported by Kenyon Ralph.  <stenn@ntp.org>
 [Bug 3456] Use uintptr_t rather than size_t to store an integer in a pointer
 - According to Brooks Davis, there was only one location <perlinger@ntp.org>
 [Bug 3449] ntpq - display "loop" instead of refid [...] <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3445] Symmetric peer won't sync on startup <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3442] Fixes for ntpdate as suggested by Gerry Garvey,
 with modifications
 New macro REFID_ISTEXT() which is also used in ntpd/ntp_control.c.
 [Bug 3434] ntpd clears STA_UNSYNC on start <perlinger@ntp.org>
 - applied patch by Miroslav Lichvar
 [Bug 3426] ntpdate.html -t default is 2 seconds.  Leonid Evdokimov.
 [Bug 3121] Drop root privileges for the forked DNS worker <perlinger@ntp.org>
 - integrated patch by  Reinhard Max
 [Bug 2821] minor build issues <perlinger@ntp.org>
 - applied patches by Christos Zoulas, including real bug fixes
 html/authopt.html: cleanup, from <stenn@ntp.org>
 ntpd/ntpd.c: DROPROOT cleanup.  <stenn@ntp.org>
 Symmetric key range is 1-65535.  Update docs.   <stenn@ntp.org>

--
NTP 4.2.8p11 (Harlan Stenn <stenn@ntp.org>, 2018/02/27)

Focus: Security, Bug fixes, enhancements.

Severity: MEDIUM

This release fixes 2 low-/medium-, 1 informational/medum-, and 2 low-severity
vulnerabilities in ntpd, one medium-severity vulernability in ntpq, and
provides 65 other non-security fixes and improvements:

* NTP Bug 3454: Unauthenticated packet can reset authenticated interleaved
	association (LOW/MED)
   Date Resolved: Stable (4.2.8p11) 27 Feb 2018
   References: Sec 3454 / CVE-2018-7185 / VU#961909
   Affects: ntp-4.2.6, up to but not including ntp-4.2.8p11.
   CVSS2: MED 4.3 (AV:N/AC:M/Au:N/C:N/I:N/A:P) This could score between
	2.9 and 6.8.
   CVSS3: LOW 3.1 CVSS:3.0/AV:N/AC:H/PR:L/UI:N/S:U/C:N/I:N/A:L This could
	score between 2.6 and 3.1
   Summary:
	The NTP Protocol allows for both non-authenticated and
	authenticated associations, in client/server, symmetric (peer),
	and several broadcast modes. In addition to the basic NTP
	operational modes, symmetric mode and broadcast servers can
	support an interleaved mode of operation. In ntp-4.2.8p4 a bug
	was inadvertently introduced into the protocol engine that
	allows a non-authenticated zero-origin (reset) packet to reset
	an authenticated interleaved peer association. If an attacker
	can send a packet with a zero-origin timestamp and the source
	IP address of the "other side" of an interleaved association,
	the 'victim' ntpd will reset its association. The attacker must
	continue sending these packets in order to maintain the
	disruption of the association. In ntp-4.0.0 thru ntp-4.2.8p6,
	interleave mode could be entered dynamically. As of ntp-4.2.8p7,
	interleaved mode must be explicitly configured/enabled.
   Mitigation:
	Implement BCP-38.
	Upgrade to 4.2.8p11, or later, from the NTP Project Download Page
	    or the NTP Public Services Project Download Page.
	If you are unable to upgrade to 4.2.8p11 or later and have
	    'peer HOST xleave' lines in your ntp.conf file, remove the
	    'xleave' option.
	Have enough sources of time.
	Properly monitor your ntpd instances.
	If ntpd stops running, auto-restart it without -g .

as/as.c
/*-
 * Copyright (c) 2012 Joseph Koshy
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <err.h>
#include <getopt.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "_elftc.h"

#include <libelftc.h>

ELFTC_VCSID("$Id: as.c 3581 2017-10-13 23:27:45Z emaste $");

enum as_long_option_index {
	AS_OPT_DEFSYM,
	AS_OPT_FATAL_WARNINGS,
	AS_OPT_LCL,
	AS_OPT_LLW,
	AS_OPT_LLW2,
	AS_OPT_LRW,
	AS_OPT_MD,
	AS_OPT_STATISTICS,
	AS_OPT_STRIP_LOCAL_ABSOLUTE,
	AS_OPT_TARGET_HELP,
	AS_OPT_VERSION,
	AS_OPT_WARN,
	AS_OPT__LAST
};

struct as_options {
	unsigned int as_listing_flags;
};

#define	AS_OPTION_SHORT_OPTIONS	":a:fghm:no:qswDI:JKLMRVWXZ"
const struct option as_option_long_options[] = {
	{ "defsym", required_argument, NULL, AS_OPT_DEFSYM },
	{ "fatal-warnings", no_argument, NULL, AS_OPT_FATAL_WARNINGS },
	{ "gen-debug", no_argument, NULL, 'g' },
	{ "help", no_argument, NULL, 'h' },
	{ "keep-locals", no_argument, NULL, 'L' },
	{ "listing-lhs-width", required_argument, NULL, AS_OPT_LLW },
	{ "listing-lhs-width2", required_argument, NULL, AS_OPT_LLW2 },
	{ "listing-rhs-width", required_argument, NULL, AS_OPT_LRW },
	{ "listing-cont-lines", required_argument, NULL, AS_OPT_LCL },
	{ "mri", no_argument, NULL, 'M' },
	{ "no-warn", no_argument, NULL, 'W' },
	{ "statistics", no_argument, NULL, AS_OPT_STATISTICS },
	{ "strip-local-absolute", no_argument, NULL,
	  AS_OPT_STRIP_LOCAL_ABSOLUTE },
	{ "target-help", no_argument, NULL, AS_OPT_TARGET_HELP },
	{ "version", no_argument, NULL, AS_OPT_VERSION },
	{ "warn", no_argument, NULL, AS_OPT_WARN },
	{ "MD", required_argument, NULL, AS_OPT_MD },
	{ NULL, 0, NULL, 0 }
};

#define	AS_OPTION_LISTING_DEFAULT "hls"

#define	AS_OPTION_USAGE_MESSAGE "\
Usage: %s [options] file...\n\
  Assemble an ELF object.\n\n\
  Options:\n\
  -D                        Print assembler debug messages.\n\
  -I DIR                    Add directory to the search list.\n\
  -J                        Suppress warnings about signed overflows.\n\
  -K                        Warn about alterations to difference tables.\n\
  -L | --keep-locals        Keep local symbols.\n\
  -R                        Merge the data and text sections.\n\
  -V                        Display the assembler version number.\n\
  -W | --no-warn            Suppress warnings.\n\
  -Z                        Generate the object even if there are errors.\n\
  -a[listing-options...]    Control assembler listings.\n\
  -g | --gen-debug          Generate debugging information.\n\
  -h | --help               Show a help message.\n\
  -march=CPU[,+EXT...]      Generate code for cpu CPU and extensions EXT.\n\
  -mtune=CPU                Optimize for cpu CPU.\n\
  -n                        Do not optimize code alignment.\n\
  -o OBJ                    Write the assembled object to file OBJ.\n\
  -q                        Suppress some warnings.\n\
  --MD FILE                 Write dependency information to FILE.\n\
  --defsym SYMBOL=VALUE     Define symbol SYMBOL with value VALUE.\n\
  --fatal-warnings          Treat warnings as fatal errors.\n\
  --listing-lhs-width=NUM   Set width of the output data column.\n\
  --listing-lhs-width2=NUM  Set the width of continuation lines.\n\
  --listing-rhs-width=NUM   Set the max width of source lines.\n\
  --listing-cont-lines=NUM  Set the maximum number of continuation lines.\n\
  --statistics              Print statistics at exit.\n\
  --strip-local-absolute    Strip local absolute symbols.\n\
  --target-help             Show target-specific help messages.\n\
  --version                 Print a version identifier and exit.\n\
  --warn                    Print warnings.\n\
  [target options]          Target specific options.\n\n\
  Options '-f', '-s', '-w', '-M', '-X' and '--mri' are accepted for\n\
  compatibility with other assemblers, but are ignored.\n"

void
as_option_usage(int iserror, const char *format, ...)
{
	va_list args;

	if (format) {
		va_start(args, format);
		vwarnx(format, args);
		va_end(args);
	}

	(void) fprintf(iserror ? stderr : stdout,
	    AS_OPTION_USAGE_MESSAGE, ELFTC_GETPROGNAME());

	exit(iserror != 0);
}

static void
as_option_listing(char *flags)
{
	(void) flags;
}

int
main(int argc, char **argv)
{
	int option, option_index;

	opterr = 0;	  /* Suppress error messages from getopt(). */

	for (option_index = -1;
	     (option = getopt_long(argc, argv, AS_OPTION_SHORT_OPTIONS,
		 as_option_long_options, &option_index)) >= 0;
	     option_index = -1)
	{
		switch (option) {

		case AS_OPT_VERSION:
			/*
			 * Print a version identifier and exit.
			 */
			(void) printf("%s (%s)\n",
			    ELFTC_GETPROGNAME(), elftc_version());
			exit(0);
			break;

		case 'h':	/* Display a help message. */
			as_option_usage(0, NULL);
			break;

		case 'f': case 's': case 'w': case 'M': case 'X':
			/*
			 * These options are accepted for compatibility
			 * reasons, but are ignored.
			 */
			break;

		case ':':

			/*
			 * A missing option argument: if the user
			 * supplied a bare '-a', supply a default set
			 * of listing control flags.
			 */
			if (optopt == 'a')
				as_option_listing(AS_OPTION_LISTING_DEFAULT);
			else
				errx(1, "option \"-%c\" expects an "
				    "argument.", optopt);
			break;

		case '?':	/* An unknown option. */
			if (optopt)
				as_option_usage(1,
				    "ERROR: unrecognized option '-%c'.",
				    optopt);
			else
				as_option_usage(1,
				    "ERROR: Unrecognized option \"--%s\".",
				    argv[optind-1]);
			break;

		default:
			if (option_index >= 0)
				errx(1,
				    "ERROR: option \"--%s\" is unimplemented.",
				    as_option_long_options[option_index].name);
			else
				errx(1,
				    "ERROR: option '-%c' is unimplemented.",
				    option);
		}
	}

	exit(0);
}
